Generate and serve default certificate for Ingress

gdemonet opened the pull request:
**Component**: salt, ingress

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**:

Whenever an `nginx-ingress-controller` configures its NGINX for an Ingress without a host or without TLS configured, it will set it up to use a _Fake Certificate_ that it generates on startup.
As such, any restart of the Pods will imply new certs being generated, which will break clients trust.

We also would like to use a CA for those certificates, allowing one to trust this CA instead of self-signed certs (this will be particularly useful for Dex setup, where API Server needs to trust the Ingress endpoint of Dex - see #2025).

**Summary**:

- Generate a CA for Ingress controllers (`/etc/metalk8s/pki/nginx-ingress/ca.[crt|key]`)
- Generate default server certs for both Ingress controllers (`/etc/metalk8s/pki/nginx-ingress/[control|workload]-plane-server.[crt|key]`)
- Store control-plane server cert/key in a _Secret_ (`metalk8s-ingress/ingress-control-plane-default-certificate`)
- Use that _Secret_ as a default certificate for the `nginx-ingress-control-plane` controller

**Acceptance criteria**: 

- When accessing a path served by the control-plane Ingress (`{bootstrap_control_plane_ip}:8443/...`), one receives a certificate with CN `nginx-ingress-control-plane-server`, issued by a CA called `ingress-ca`
- After restarting the `nginx-ingress-control-plane` controller Pod(s), one can refresh the UI without the need to accept new certificates

gdemonet committed:
salt: Setup CA for nginx-ingress

In order to sign server certificates for both of our
`nginx-ingress-controller`s, we need a CA, which we add to the CAs
managed by the `ca:minion`.

gdemonet committed:
salt: Generate cert for workload-plane Ingress

We add formulas to generate a server certificate for workload-plane
Ingress controller to use as a default certificate.

gdemonet committed:
salt: Generate cert for control-plane Ingress

We add formulas to generate a server certificate for control-plane
Ingress controller.

gdemonet committed:
salt: Store control-plane Ingress cert as Secret

Note that we ensure the certs are available on the Node where
salt-master runs (although it seems unproper to do it this way, the only
explicit requirement about those certs is to have them available in the
Salt master container).

gdemonet committed:
chart: Use default certificate for control-plane Ingress

In order to trust the default certificate used by all of our
control-plane Ingress objects that don't specify a host, we need to
rely on the default certificate.

This will enable API Server to validate an OIDC provider if it was
served behind Ingress.

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

gdemonet commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort










*The following options are set:* **approve**

TeddyAndrieux commented on salt/metalk8s/addons/nginx-ingress-control-plane/deployed/tls-secret.sls:
```diff
@@ -0,0 +1,19 @@
+#!jinja | kubernetes kubeconfig=/etc/kubernetes/admin.conf&context=kubernetes-admin@kubernetes
```
```suggestion
#!jinja | metalk8s_kubernetes kubeconfig=/etc/kubernetes/admin.conf&context=kubernetes-admin@kubernetes
```

TeddyAndrieux commented on salt/metalk8s/addons/nginx-ingress/certs/server.sls:
```diff
@@ -0,0 +1,45 @@
+
+Generate Workload-Plane Ingress server certificate:
+  x509.certificate_managed:
+    - name: /etc/metalk8s/pki/nginx-ingress/workload-plane-server.crt
```
Why generate this cert if we not use it, we should use it as default cert for `nginx-ingress` even if it's not usefull for the moment

gdemonet replied:
I'll do it :)

TeddyAndrieux reviewed (changes requested):
Since you add a new CA you need to add it to the backup and restore scripts

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux







*The following options are set:* **approve**

Ebaneck commented:
> Since you add a new CA you need to add it to the backup and restore scripts

Lets track it here since they are similar https://github.com/scality/metalk8s/issues/2031

TeddyAndrieux commented:
> > Since you add a new CA you need to add it to the backup and restore scripts
> 
> Lets track it here since they are similar #2031

Depends when @alexandre-allard-scality PR about backup/restore test will be merge, the backup ticket need to be tackled before backup/restore get merged

gdemonet commented:
> Since you add a new CA you need to add it to the backup and restore scripts

OK I'll add this. Though won't be able to push changes before tonight.

Ebaneck commented:
> > Since you add a new CA you need to add it to the backup and restore scripts
> 
> OK I'll add this. Though won't be able to push changes before tonight.

Never mind... i am going to deal with this

Ebaneck committed:
chart: Use default certificate for workload-plane Ingress

In order to trust the default certificate used by all of our
workload-plane Ingress objects that don't specify a host, we need to
rely on the default certificate.

Ebaneck committed:
salt: Store workload-plane Ingress cert as Secret

Note that we ensure the certs are available on the Node where
salt-master runs (although it seems unproper to do it this way, the only
explicit requirement about those certs is to have them available in the
Salt master container).

TeddyAndrieux commented on scripts/downgrade.sh.in:
```diff
@@ -107,13 +107,23 @@ downgrade_bootstrap () {
         'repositories': $repo_endpoint}}}" \
         --retcode-passthrough
 
-    SALT_MASTER_CALL=(crictl exec -i "$(get_salt_container)")
```
Do not remove it as the master container ID change after the `roles.bootstrap`

Ebaneck replied:
ok

Ebaneck committed:
salt: Ensure CA files are available

We need to ensure that CA files are available while testing upgrade
and downgrade from one version to the other.
To enforce this, we run the Ca role during bootstrap to create these
CA certificates.

TeddyAndrieux reviewed (approved):
LGTM

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye gdemonet.
